api/src/nv_ingest_api/util/dataloader/dataloader.py

```diff
@@ -0,0 +1,177 @@
+                (
+                    ffmpeg.input(
+                        "pipe:",
+                    )
+                    .output(str(output_pattern), f="segment", segment_time=segment_time, c="copy", map="0")
+                    .run(capture_stdout=True, capture_stderr=True, input=f_in.read())
+                )
+        except ffmpeg.Error as e:
+            logging.error(f"FFmpeg error for file {input_path}: {e.stderr.decode()}")
+        files = [str(output_dir / f"{file_name}_chunk_{i:04d}{suffix}") for i in range(int(num_splits))]
+        return files
+
+
+def process_data(
+    queue: queue.Queue,
+    interface: LoaderInterface,
+    paths: list[str],
+    output_dir: str,
+    thread_stop: threading.Event,
+    split_interval: int = 450,
+    files_completed: list[str] = None,
+):
+    for path in paths:
+        if thread_stop:
+            return
+        # check for file size greater than 450MB
+        files = []
+        if path.stat().st_size > (split_interval * 1e6):
+            files = interface.split(path, output_dir, split_interval)
+        # either path or files need to be read and pushed to queue
+        else:
+            files = [path]
+        for file in files:
+            if thread_stop:
+                return
+            with open(file, "rb") as f:
+                queue.put(f.read())
+        files_completed.extend(files)
+    queue.put(StopIteration)
+    return files_completed
+
+
+class DataLoader:
+    def __init__(
+        self,
+        paths: list[str],
+        output_dir: str,
+        split_interval: int = 450,
+        interface: LoaderInterface = None,
+        size: int = 2,
+    ):
+        self.thread = None
+        self.thread_stop = False
+        self.queue = queue.Queue(size)
+        self.paths = [Path(path) for path in paths]
+        self.output_dir = output_dir
+        self.split_interval = split_interval
+        self.chunk_count = 0
+        self.interface = interface
+        self.files_completed = []
+
+    def __getitem__(self, index: int):
+        return self.__next__()
+
+    def __next__(self):
+        payload = self.queue.get()
+        if payload == StopIteration:
+            raise payload
+        else:
+            return payload
+
+    def stop(self):
+        if self.thread:
+            self.thread_stop = True
+            self.thread.join()
+        self.thread_stop = False
+        while self.queue.qsize() != 0:
+            with self.queue.mutex:
+                self.queue.queue.clear()
+
+    def __iter__(self):
+        self.stop()
+        self.files_completed = []
+        self.thread = threading.Thread(
+            target=process_data,
+            args=(
+                self.queue,
+                self.interface,
+                self.paths,
+                self.output_dir,
+                self.thread_stop,
+                self.split_interval,
+                self.files_completed,
+            ),
+            daemon=True,
```

Comment:
This should be more RAII-like. Could make a best erffort cleanup in `__del__`  and/or `__exit__`  , but we shouldn't rely on the user to know they have to call stop or they'll leak a thread instance.

Reply:
The user shouldn't every call stop. This is something we use to reset the iterator. We want to be able to continue using the same iterator (from where it stopped) if a user decides to break during an iteration and then still access the rest of the data using next(itr). But I can definitely add a del and exit that does the same.